tiledb/sm/serialization/query.cc

```diff
@@ -1317,22 +1498,80 @@ Status query_from_capnp(
   // on the reader or writer directly Now we set it on the query class after the
   // heterogeneous coordinate changes
   if (type == QueryType::READ) {
-    auto reader_reader = query_reader.getReader();
-    auto reader = (Reader*)query->strategy();
+    bool found = false;
+    bool use_refactored_readers = false;
+    RETURN_NOT_OK(query->config()->get<bool>(
+        "sm.use_refactored_readers", &use_refactored_readers, &found));
+    assert(found);
+    if (use_refactored_readers && !schema->dense() &&
```

Comment:
Instead of relying on the config parameter we should check to see if the new reader state is set or not, `query_reader.hasReaderIndex()` should be effective here.